apps/req.c

```diff
@@ -837,14 +837,15 @@ int req_main(int argc, char **argv)
             /* Set up V3 context struct */
             X509V3_set_ctx(&ext_ctx, CAcert != NULL ? CAcert : new_x509,
                            new_x509, NULL, NULL, X509V3_CTX_REPLACE);
-            if (CAcert == NULL) { /* self-issued, possibly self-signed */
-                if (!X509V3_set_issuer_pkey(&ext_ctx, pkey)) /* prepare right AKID */
-                    goto end;
-                ERR_set_mark();
-                if (!X509_check_private_key(new_x509, pkey))
+            /* prepare fallback for AKID, but only if self-signed */
+            if (CAcert == NULL) {
+                if (cert_matches_key(new_x509, pkey)) {
+                    if (!X509V3_set_issuer_pkey(&ext_ctx, pkey))
+                        goto end;
+                } else {
```

Comment:
Uhm, doesn't this *take away* the possibility of a self-issued but not self-signed cert?

Reply:
No, it just precludes the generation of a (wrong) fallback AKID value for such certs.
See `v3_akid.c`:
```
        if (keyid == 2 && ikeyid == NULL && ctx->issuer_pkey != NULL) {
            /* generate fallback AKID, emulating s2i_skey_id(..., "hash") */
```
The `"Warning: Signature key and public key of cert do not match\n"` is not an error.

Reply:
Aha ok